src/abstract_domains.rs

```diff
@@ -126,9 +194,67 @@ pub enum ExpressionDomain {
         // The value of this expression if join_condition is false.
         alternate: Box<ExpressionDomain>,
     },
+
+    // An expression that is true if left and right are equal.
+    Equals {
+        // The value of the left operand.
+        left: Box<ExpressionDomain>,
+        // The value of the right operand.
+        right: Box<ExpressionDomain>,
+    },
+
+    /// An expression that is true if the operand is false.
+    Not { operand: Box<ExpressionDomain> },
+
+    // An expression that is true if either one of left or right are true.
+    Or {
+        // The value of the left operand.
+        left: Box<ExpressionDomain>,
+        // The value of the right operand.
+        right: Box<ExpressionDomain>,
+    },
 }
 
 impl AbstractDomain for ExpressionDomain {
+    /// Returns an expression that is "self && other".
+    fn and(&self, other: &Self) -> Self {
+        if self.must_be_true() {
+            if other.must_be_true() {
+                ExpressionDomain::CompileTimeConstant(ConstantValue::True)
+            } else {
+                other.clone()
+            }
+        } else if other.must_be_true() {
+            self.clone()
+        } else {
+            // todo: #32 more simplifications
+            ExpressionDomain::And {
+                left: box self.clone(),
+                right: box other.clone(),
+            }
+        }
```

Comment:
match (self.must_be_true(), other.must_be_true()) {
    (True, True) => ExpressionDomain::CompileTimeConstant(ConstantValue::True),
    (True, False) => other.clone(),
    (False, True) => self.clone(),
    (False, False) => ExpressionDomain::And { ... }

Makes it easier to see the logic table for must_be_trues(). But if you plan on adding a lot more than that it may or may not be helpful, up to you.

Reply:
I don't like the looks of the (False, False) option because it obscures the 3-valued logic behavior of must_be_true. I have it in mind to make must_be_true return Option<bool> to make it more painfully obvious that !must_be_true does not imply must_be_false. When I do that, I'll have another look at the match constructor for this.